xcdat/averager.py

```diff
@@ -0,0 +1,501 @@
+            # TODO: we may want to attach the bounds to the dimension weights, too
+            dim_wts.attrs = dim_bounds.attrs
+            dim_wts.name = dim_name + "_wts"
+            weights[dim_name] = dim_wts
+
+    return weights
+
+
+def _get_uniform_weights(xvariable: xr.DataArray) -> AxisWeights:
+    """Gets uniform weights for a dimension with unweighted bounds.
+
+    Weights are uniform (filled with ones) based on the shape of the data's
+    dimension.
+
+    Parameters
+    ----------
+    xvariable : xr.DataArray
+        xr.DataArray produced using ``open_variable``.
+
+    Returns
+    -------
+    AxisWeights
+        Weights filled with ones.
+    """
+    weights: AxisWeights = {}
+
+    for dim in xvariable.dims:
+        xdim = xvariable[dim]
+        dim_wts = xr.DataArray(
+            name=f"{dim}_wts",
+            data=np.ones(len(xdim)),
+            dims=xdim.dims,
+            coords=xdim.coords,
+        )
+        weights[dim] = dim_wts
+    return weights
+
+
+def _scale_dim_to_region(
+    dimension: xr.DataArray, region_bounds: List[float]
+) -> xr.DataArray:
+    """Scales down a dimension's bounds based on regional selection.
+
+    Parameters
+    ----------
+    bounds : xr.DataArray
+        The entire dimension's bounds.
+    region_bounds : List[float]
+        The dimension's regional bounds to adjust the entire dimension bounds to.
+
+    Returns
+    -------
+    xr.DataArray
+        Scaled dimension bounds based on regional selection.
+    """
+    # Adjustments for above / right of region
+    dimension[dimension[:, 0] > np.max(region_bounds), 0] = np.max(region_bounds)
+    dimension[dimension[:, 1] > np.max(region_bounds), 1] = np.max(region_bounds)
+    # Adjustments for below / left of region
+    dimension[dimension[:, 0] < np.min(region_bounds), 0] = np.min(region_bounds)
+    dimension[dimension[:, 1] < np.min(region_bounds), 1] = np.min(region_bounds)
+    return dimension
+
+
+def _combine_axis_weights(
+    axis_weights: AxisWeights, axis: List[SupportedAxes]
+) -> xr.DataArray:
+    """Generically rescales axis weights for a given region.
+
+    Operations include:
+
+    - Create N-dimensional weighting array based on axis_weights and axis
+       dimensions specified
+
+    Parameters
+    ----------
+    axis_weights : AxisWeights
+        Dictionary of axis weights (using dims as keys).
+    axis : List[SupportedAxes]
+        List of axes that should be weighted.
+
+    Returns
+    -------
+    xr.DataArray
+        Array of weights
+    """
+    # FIXME: This assumes the axis weights dict keys are in the same order as the axis list.
+    weights = axis_weights[axis[0]]
+    # FIXME: By setting the initial value for weights to the first element (axes),
+    # it will multiple with itself in the loop below. Is this the intended behavior?
+    for axes in axis:
+        weights = weights * axis_weights[axes]
+    return weights
+
+
+def _averager(
```

Comment:
Hmmmm, this will be something good to take a closer look at. I'm not sure why CDAT performs poorly with larger domains. I'd have to dig into the algorithm to come up with an idea.

Also, it might help to use `timeit.repeat` to perform repeated runs to improve accuracy, unless that's what you already did. If we can leverage xarray `_xraverage()`, it would simplify our testing and implementation significantly!